kpasswd/kpasswdd.c

```diff
@@ -691,48 +717,67 @@ doit(krb5_keytab keytab, int port)
     struct sockaddr_storage __ss;
     struct sockaddr *sa = (struct sockaddr *)&__ss;
-
-    if (explicit_addresses.len) {
-	addrs = explicit_addresses;
+    int fdz;
+    int from_inetd;
+    socklen_t fromlen;
+    krb5_address my_addr;
+    struct sockaddr_storage __local;
+    struct sockaddr *localsa = (struct sockaddr *)&__local;
+
+    fromlen = sizeof __ss;
+    from_inetd = (getsockname(0, sa, &fromlen) == 0);
+
+    if (!from_inetd) {
+        if (explicit_addresses.len) {
+            addrs = explicit_addresses;
+        } else {
+            ret = krb5_get_all_server_addrs(context, &addrs);
+            if (ret)
+                krb5_err(context, 1, ret, "krb5_get_all_server_addrs");
+        }
+        n = addrs.len;
+
+        sockets = malloc(n * sizeof(*sockets));
+        if (sockets == NULL)
+            krb5_errx(context, 1, "out of memory");
+        maxfd = -1;
+        FD_ZERO(&real_fdset);
+        for (i = 0; i < n; ++i) {
+            krb5_socklen_t sa_size = sizeof(__ss);
+
+            krb5_addr2sockaddr(context, &addrs.val[i], sa, &sa_size, port);
+
+            sockets[i] = socket(__ss.ss_family, SOCK_DGRAM, 0);
+            if (sockets[i] < 0)
+                krb5_err(context, 1, errno, "socket");
+            if (bind(sockets[i], sa, sa_size) < 0) {
+                char str[128];
+                size_t len;
+                int save_errno = errno;
+
+                ret = krb5_print_address(&addrs.val[i], str, sizeof(str), &len);
+                if (ret)
+                    strlcpy(str, "unknown address", sizeof(str));
+                krb5_warn(context, save_errno, "bind(%s)", str);
+                continue;
+            }
+            maxfd = max(maxfd, sockets[i]);
+            if (maxfd >= FD_SETSIZE)
+                krb5_errx(context, 1, "fd too large");
+            FD_SET(sockets[i], &real_fdset);
+        }
     } else {
-	ret = krb5_get_all_server_addrs(context, &addrs);
-	if (ret)
-	    krb5_err(context, 1, ret, "krb5_get_all_server_addrs");
-    }
-    n = addrs.len;
-
-    sockets = malloc(n * sizeof(*sockets));
-    if (sockets == NULL)
-	krb5_errx(context, 1, "out of memory");
-    maxfd = -1;
-    FD_ZERO(&real_fdset);
-    for (i = 0; i < n; ++i) {
-	krb5_socklen_t sa_size = sizeof(__ss);
-
-	krb5_addr2sockaddr(context, &addrs.val[i], sa, &sa_size, port);
-
-	sockets[i] = socket(__ss.ss_family, SOCK_DGRAM, 0);
-	if (sockets[i] < 0)
-	    krb5_err(context, 1, errno, "socket");
-	if (bind(sockets[i], sa, sa_size) < 0) {
-	    char str[128];
-	    size_t len;
-	    int save_errno = errno;
-
-	    ret = krb5_print_address(&addrs.val[i], str, sizeof(str), &len);
-	    if (ret)
-		strlcpy(str, "unknown address", sizeof(str));
-	    krb5_warn(context, save_errno, "bind(%s)", str);
-	    continue;
-	}
-	maxfd = max(maxfd, sockets[i]);
-	if (maxfd >= FD_SETSIZE)
-	    krb5_errx(context, 1, "fd too large");
-	FD_SET(sockets[i], &real_fdset);
+        n = 1;
+        maxfd = 0;
+	fdz = 0;
```

Comment:
I'm not sure that this is actually portable.  We are using stdin assuming that it works as input and output later in the program.
I'm also not comfortable that we don't dup2(2) it up to being larger than two as well as opening /dev/null and dup2(2) the resulting fd to 0, 1, and 2.